### ISSUE TITLE
Feature Request: fish_kak_keybindings

### ISSUE THREAD

**Issue by JacobTravers** [fish-shell/fish-shell#7748]
I would like to contribute to create fish_kak_keybindings which would use kakoune's modal editing and bindings.

Changing keybindings to match kakoune's seems like it should be simple enough. However I'm wondering if anybody might know where to start to make `escape` go directly to visual mode, in kakoune, moving is selecting so `<w>` does not just go to the next word, it selects from current position to the next word. Also by convention, capital commands expand selection, so `<W>` expands selection to the next word. Abstractly i think an easy way to do it would be:

`<shift-{key}>` continue visual mode -> movement
`<key>` reset visual-mode -> movement

Also not sure how using `<i>` to insert before and `<a>`  to insert after selection would work or how hard it may be.

If anybody has any info or is willing to help with this I would appreciate it very much!

**Comment by krobelus** [comment]
I think it's doable, but just like with vi mode, some features are not yet implemented.
Also, you can use <kbd>Alt+E</kbd> or <kbd>Alt+V</kbd> to use your real editor.

This example hacks vi mode to go straight to visual mode on <kbd>Escape</kbd> and makes <kbd>b</kbd> behave as you describe

```diff
diff --git a/share/functions/fish_vi_key_bindings.fish b/share/functions/fish_vi_key_bindings.fish
index 3dfc6dbca..c745aa82d 100644
--- a/share/functions/fish_vi_key_bindings.fish
+++ b/share/functions/fish_vi_key_bindings.fish
@@ -74 +74 @@ function fish_vi_key_bindings --description 'vi-like key bindings for fish'
-    bind -s --preset -M insert \e "if commandline -P; commandline -f cancel; else; set fish_bind_mode default; commandline -f backward-char repaint-mode; end"
+    bind -s --preset -M insert \e "if commandline -P; commandline -f cancel; else; commandline -f begin-selection; set fish_bind_mode visual; commandline -f backward-char repaint-mode; end"
@@ -278,2 +278,4 @@ function fish_vi_key_bindings --description 'vi-like key bindings for fish'
-    bind -s --preset -M visual b backward-word
-    bind -s --preset -M visual B backward-bigword
+    bind -s --preset -M visual b begin-selection backward-word
+    bind -s --preset -M visual B backward-word
+    bind -s --preset -M visual \eb begin-selection backward-bigword
+    bind -s --preset -M visual \eB backward-bigword
```

Of course the final implementation should not change vi-mode.

**Comment by EPonyA** [comment]
"Kakoune everywhere" is my dream. I would so love this to be supported.

**Comment by sshilovsky** [comment]
I would love to see helix mode too. Helix is highly inspired by kakoune, and I believe the differences would be only in default keybinding choices.

Do you think there should be a separate mode for it then, like `fish_hx_keybindings`?

**Comment by sshilovsky** [comment]
Is there a way to write tests for keybindings?

**Comment by sshilovsky** [comment]
Both kakoune and helix seem to have specific behavior for w/b/e motions, different from vim. I don't think this behavior can be programmed with existing `bind` commands. Would you accept a patch with additional commands like `helix-forward-word`?

**Comment by ismay** [comment]
> Both kakoune and helix seem to have specific behavior for `w`/`b`/`e` motions, different from vim. I don't think this behavior can be programmed with existing `bind` commands. Would you accept a patch with additional commands like `helix-forward-word`?
> 
> The difference from vim here comes from the fact that kakoune and helix use inclusive ranges every time, e.g. the cursor is by design a selection of length `1`, unlike in vim, where it's a selection of length `0`. So moving to next word is less straightforward.
> 
> Same applies to `t` and `T` motions.

Yeah I agree. It seems like in order to be able to implement this we'd need some new special input functions. I don't think they'd necessarily have to be coupled to an editor (i.e. helix-forward-word), they'd just have to be more specific than just `forward-word`. Besides allowing for helix/kakoune style bindings, that could also allow bindings to be more expressive.

**Comment by 0x4c756e61** [comment]
Any further development on this? I am highly interested into using the helix keybinds in fish

**Comment by krobelus** [comment]
there's an impressive helix key bindings plugin but it does
hit some limitations that'd need some changes in fish

**Comment by jtrv** [comment]
> Any further development on this? I am highly interested into using the helix keybinds in fish

I gave up on this a long time ago in favor of just using the edit-prompt keybinding as described by krobelus:

> Also, you can use Alt+E or Alt+V to use your real editor.

Much simpler with the added bonus of having all your configs / lsp, etc

**Comment by rouge8** [comment]
The helix bindings at https://git.ersei.net/nix-configs.git/tree/home/common/shell/functions are pretty cool too!

**Comment by sshilovsky** [comment]
I appreciate references to the original repo for the helix bindings: https://github.com/sshilovsky/fish-helix

PRs are welcome too

**Comment by sabisho** [comment]
Now that [Nushell](https://github.com/nushell/reedline/issues/639#issuecomment-4620324501) is adding a Helix mode, it would be great if Fish also added Helix keybindings natively to its core, rather than relying on a plugin that might not offer the polished modal editing experience users expect.

**Comment by nettlesh** [comment]
I also strongly agree that I would love native Helix bindings. Other shells are implementing them as first-class citizens. The biggest reason to make them native vs a plugin in my eyes is that it encourages CLI tool developers to add support for those bindings out of the box. If no one else is actively working on it, I'm happy to take a crack at it.
